### PR TITLE
2018 Prototype

### PR DESCRIPTION
**Pull request by lederer**
This PR adds a new prototype that uses DeviceOrientation to guide the user through orienting the device to the correct bearing within each discovery portal. 

See `/prototype/README.md` for details.

### CODE REVIEW

File: prototype/.babelrc
Comment (kellyi) [inline prototype/.babelrc]: Mentioned this via Slack but it turns out the order in which these presets are listed is significant, so I believe this needs to be `["env", "react", "stage-2"]` to avoid an error related to not being able to parse the class-properties method syntax.

Comment (lederer) [reply prototype/.babelrc]: Did it work? I tried reordering those on an unrelated project with the same problem and it didn't help.

Comment (kellyi) [reply prototype/.babelrc]: Ooops, I lied: we may need to explicitly include `babel-plugin-transform-class-properties`. I'll push a commit with this.

**Comment by kellyi** [comment]
Tried this out and it's working well! As far as I'm concerned, this is good to merge in -- or keep open, if we'd like to continue working on the prototype version. Going to make another quick glance through the code in case there's anything worth pointing out. From what I can tell so far, having tested this out on iOS/iPhone 6S:

- the device orientation interface stuff works very well on iOS. Since the values that API sends out are rapid and incessant until turned off, the production version of this will need to adopt this prototype's debounce smoothing

- the geolocation works well-ish, with an asterisk which is: the level of granularity the zones seem to expect is within ten feet or so, but the geolocation API doesn't seem to provide values that accurate. We can try tweaking this or doing some work to complement the geolocation data with some additional calculations to try to make the reading more accurate, but at base I think we're going to have to supplement (or supplant!) the geofence feature with a way to manually select zones. In fact, it might be worth having the manual zone selection rather than the geofence be the default way this works, but we can figure that out as we go.

Will write this up in the form of a brief ADR and open a PR with that. Generally speaking, though, this looks really great!

File: prototype/src/js/AlphaOrienter.jsx
Comment (kellyi) [inline prototype/src/js/AlphaOrienter.jsx:80]: Seems like this `classnames` library might be useful to include in other Civic Apps React projects.

File: prototype/src/js/DeviceOrienter.jsx
Comment (kellyi) [inline prototype/src/js/DeviceOrienter.jsx:89]: Is there a difference in scope or something between the `onDeviceOrientation = event =>` style method declarations and the ones like `removeOrientationListener()`?

File: prototype/src/js/DeviceOrienter.jsx
Comment (kellyi) [inline prototype/src/js/DeviceOrienter.jsx:22]: I hacked together a tiny prototype that just logs the values out and routes them through Redux state rather than React state. Both options seem to work pretty well; the Redux version looks something like:

```js
const deviceOrientation = 'deviceorientation';

function startFindOrientation() {
    return {
        type: START_FIND_ORIENTATION,
    };
}

function completeFindOrientation({
    absolute,
    alpha,
    beta,
    gamma,
}) {
    return (dispatch) => {
        dispatch({
            type: COMPLETE_FIND_ORIENTATION,
            payload: {
                absolute,
                alpha,
                beta,
                gamma,
            },
        });
    };
}

export function stopFindOrientation() {
    return (dispatch) => {
        window
            .removeEventListener(
                deviceOrientation,
                e => dispatch(completeFindOrientation(e)),
            );
        dispatch({
            type: FAIL_FIND_ORIENTATION,
        });
    };
}

export function findOrientation() {
    return (dispatch) => {
        dispatch(startFindOrientation());
        window
            .addEventListener(
                deviceOrientation,
                e => dispatch(completeFindOrientation(e)),
            );
    };
}
```

**Comment by lederer** [comment]
> the production version of this will need to adopt this prototype's debounce smoothing

I don't think the prototype is debouncing device orientation events. Tho it is debouncing geolocation events.

File: prototype/src/js/AlphaOrienter.jsx
Comment (lederer) [reply prototype/src/js/AlphaOrienter.jsx:80]: Super useful.

File: prototype/src/js/ZoneTour.jsx
Comment (kellyi) [inline prototype/src/js/ZoneTour.jsx:6]: Probably not worth changing but for components without declared lifecycle methods we typically use functional components and destructure their props in the arguments list, like:

```
export default function ZoneTour({
    zone: {
        tour: [
            {
                image,
                caption,
            },
        ],
...etc
```

File: prototype/src/js/worlds/office.js
Comment (kellyi) [inline prototype/src/js/worlds/office.js:82]: Nitpicky but this function could also return in each of these case statements instead of assigning and then `break`ing.

File: prototype/src/js/DeviceOrienter.jsx
Comment (lederer) [reply prototype/src/js/DeviceOrienter.jsx:89]: No scope difference per se. 

I used `onDeviceOrientation = event =>` only for methods that need to be used as event handlers on jsx elements or components. It ensures `this` stays bound to the correct object (instead of having to put all those `.bind(this)` calls inside the `constructor`).

For consistency's sake, it would've been fine to make all the methods use the above syntax. Or OTOH I suppose one could argue that the syntax difference is a visual signal of which methods are intended to be true event handlers. Truth is, I hadn't thought through it.

**Review by kellyi — APPROVED** [review]
:+1: This looks great! We'll have to think through some of the geofencing details as we start to work on this but the device orientation feature -- and its implementation here -- seems quite viable!

**Comment by kellyi** [comment]
> I don't think the prototype is debouncing device orientation events.

Ah, ok, I misread. In that case I don't know that we do need to debounce orientation -- it looks a little alarming when you just see the raw numbers updating quickly but if that's not affecting the app functionality then we may be able to do without... with the caveat that we should probably make sure that both the orientation and geolocation updates to app state are sequestered well enough to prevent tangentially related unnecessary component re-renders.

File: prototype/src/js/DeviceOrienter.jsx
Comment (kellyi) [reply prototype/src/js/DeviceOrienter.jsx:89]: I like it both for the visual indicator and to avoid the `.bind(this)` statements.

File: prototype/src/js/ZoneTour.jsx
Comment (lederer) [reply prototype/src/js/ZoneTour.jsx:6]: OK cool. Will leave as is, since it's just a prototype. But will remember for the future. Lesson learned!

File: prototype/src/js/worlds/office.js
Comment (lederer) [reply prototype/src/js/worlds/office.js:82]: I've always had a resistance to multiple return statements, but perhaps it's time to let that go. Will keep in mind.

**Comment by lederer** [comment]
> it looks a little alarming when you just see the raw numbers updating quickly

Oh totally. Less so when you truncate most or all of the decimal digits. :)

What are your thoughts on merging vs leaving as a branch? Whatever is best for the project.

**Comment by kellyi** [comment]
> What are your thoughts on merging vs leaving as a branch? Whatever is best for the project.

Good q -- I'm inclined to merge. @ajrobbins what do you think about merging vs leaving open as a branch?

**Comment by kellyi** [comment]
Spoke with @ajrobbins and we decided to add a `develop` branch to this project, so I just made one. Can you retarget this PR at `develop` before merging?

**Comment by mmcfarland** [comment]
> we decided to add a develop branch to this project

I would recommend we don't target this repo for the new work.  I don't think the next version will benefit from inheriting the commit history and activity of this GH repo - it's a complete rewrite.  I'd prefer we create a new repo for this project and let this repo remain an artifact of the work we have completed.  Ideally, the ADR and the prototype would be committed to the new repo in the usual manner.

**Comment by lederer** [comment]
@kellyi Per @mmcfarland's recommendation, is there a new repo I should move the prototype to, instead of merging it into this one?

**Comment by ajrobbins** [comment]
@lederer We will make one next sprint when we set up the new project from our app template.

**Comment by caseycesari** [comment]
@lederer Created a new repo at https://github.com/azavea/pwd-urban-water-revealed

**Comment by lederer** [comment]
Great. I'll open a PR over there with the prototypes.

**Comment by lederer** [comment]
Closing this PR in favor of the new repo.

**Pull request closed without merging**
